### ISSUE TITLE
Retrieving "from" and "to" from histograms

### ISSUE THREAD

**Issue by bakura10** [elastic/elasticsearch#9528]
Hi,

I'm using the date_histogram but I realize I often need to have the "from" and "to" date in the response to build my results. Of course I could use the date_range, but the date_histogram is easier to use and avoid me to manually create the different date_ranges.

Is there a way to retrieve this information in the answer, in addition to the "key" ?

Thanks

**Comment by clintongormley** [comment]
Hi @bakura10 

No, you can't directly get the `from` and `to`, but the `key` is the `from` value, and the next `key` is the `to` value (`to` is exclusive, while `from` is inclusive).  So you're only missing the last `to` value really (which you could calculate from the `interval`).

Is this a common requirement?  (Leaving this open for others to chime in)

**Comment by bakura10** [comment]
Obviously I could compute it myself from the interval or get the next value but this is extra work to do on the server.  I actually need to explicitly show the various ranges in my charts, so having the from and to is quite a necessity.

**Comment by bleskes** [comment]
we discussed this a bit and we found it hard to find a way to expose the to in the bucket in a way that's useful, meaningful, and consistent in all the related aggs (histogram, date histogram and ranges). On the other hand, I think this problem and potentially many others will be solved if we allow for scripted key formatting. In this way you can construct any description that useful for your use case.

**Comment by bleskes** [comment]
@bakura10 would that be useful to you?

**Comment by bakura10** [comment]
Hhhmm I'm not sure that would be useful in my specific case, because what I need is only the from and to, and this is known by the histogram only.

I don't know the technical reasons, but actually having the start and end value of histogram would be quite coherent with how date range aggregation work. Actually, histogram is nothing more than a range aggregation but with automatically computed ranges. And I was quite surprised to see the format was radically different to ranges aggregations. Ideally, we should have the key_as_string, to, to_as_string, from, from_as_string, and the ability to have keyed key (like in ranges).

Now, if this is not possible, I don't think I would benefit the scripted key formatting (at least in my use case).

**Comment by bleskes** [comment]
@bakura10 the problem with the to key in the date case is the you probably don't want it to be exclusive. Think of buckets by week. When display the chart, you will say "1 jan to 6 jan" on one bar and "7 jan to 13 jan" on the second. This is not the case when doing histograms on price ranges (doubles) and also not in the ranges where the end is specified to be exclusive. We felt that a script would allow you to construct the key_as_string as is the best for your case.

**Comment by bakura10** [comment]
Hi,

The example of week is very good. Actually, in my use case, I allow people to select both start date and end date AND an interval. For instance, if people select "2nd february to 4th february" with "week" interval, Elasticsearch will actually create a bucket for a full week (so 2nd February inclusive to 9th exclusive), which is EXACTLY what I want.

In my chart, I will therefore have a data point at "2nd february" that contains aggregations from 2nd february to 9th february exclusive. The problem is that because user has explicitly selected 2nd february to 4th february", it may not be obvious what is included in the data point and what is not.

Therefore, I'd like to have a tooltip over each data point, to explicitly write in plain English what is covered (for instance: "From 2nd February to 9th February exclusive"). But for knowing that, I need the "from" and "to" that Elasticsearch has found for me based on the interval.

That's why I do not know how script for key could help, because ultimately, the bounds are calculated by Elasticsearch histogram.

Does it make sense?

**Comment by bleskes** [comment]
> From 2nd February to 9th February exclusive

I think scripted formatting (with access to the right data) will allow you to format that message exactly as you need it , right?

**Comment by bakura10** [comment]
The qustion is "how access the right data"? If elasticcsearch provides context value (the from and to), and that is available to the scripted through a "ctx" array, yes that would be indeed possilbe :).

**Comment by markharwood** [comment]
Related: https://github.com/elastic/elasticsearch/issues/28479 where we agreed to add from and to for each bucket.

cc @elastic/es-search-aggs

**Comment by colings86** [comment]
Closing in favour of https://github.com/elastic/elasticsearch/issues/28479
